# Tidepool Hackathon: General EDA

**Difference between type1 and type 2 diabetes**
Both types of diabetes are chronic diseases that affect the way your body regulates blood sugar, or glucose. Glucose is the fuel that feeds your body’s cells, but to enter your cells it needs a key. Insulin is that key.

Type 1: Do not produce insulin 
Type 2: Don't respond well enough to insulin as they should, and eventually do not produce enough insulin

Other Facts: 
- Symptoms are similar, however: 
    - Many ppl with Type 2 won't have symptoms for several years
    - Type 2 symptoms develop slowly over the course of time 
    - Type 1 develops quickly, usually in childhood and teenagehood, but can be developed later in life.
    

# Data dictionary 
    diagnosisAge: donor's diagnosis age
    yearsLivingWithDiabetes:	the number of years the donor has been living with diabetes
    basal.count	the number of basal insulin records the donor has
    bolus.count	the number of bolus insulin records
    cbg.count	the number of contiguous blood glucose records
    deviceEvent.count	the number of device activity records
    pumpSettings.count	the number of pump setting records
    smbg.count	the number of blood glucose meter records
    upload.count	the number of upload records
    wizard.count	the number of bolus calculator records
    cbg.nUniqueDeviceIds	the number of unique devices the donor has for this data type
    bolus.nUniqueDeviceIds	the number of unique devices the donor has for this data type
    smbg.nUniqueDeviceIds	the number of unique devices the donor has for this data type
    basal.nDaysWithData	the number of days the donor has this data type
    bolus.nDaysWithData	the number of days the donor has this data type
    cbg.nDaysWithData	the number of days the donor has this data type
    deviceEvent.nDaysWithData	the number of days the donor has this data type
    pumpSettings.nDaysWithData	the number of days the donor has this data type
    smbg.nDaysWithData	the number of days the donor has this data type
    upload.nDaysWithData	the number of days the donor has this data type
    wizard.nDaysWithData	the number of days the donor has this data type
    cgmSettings.count	the number of contiguous glucose monitor setting records
    cgmSettings.nDaysWithData	the number of days the donor has this data type


In [24]:
# Pulling necessary imports
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns

In [2]:
data = pd.read_csv('./type1-and-type2-data-cleaned.csv')

In [8]:
data.head(20)

hashID diagnosisType  \
0   92c2847d30b6514ac8a6385f04b3bf3c92ee87a8b89cb8...         type2   
1   6b720ecfc94a90f7c75a39f80a97eac27969ca95a2e6fb...         type2   
2   e0ef840bd07fc79f10b8e21a7f6f3d53ef3b7594d94ec0...         type1   
3   8974377383af45404d560543b58749b240c78885ca0710...         type1   
4   f92417af60cc16f254600f310df4d15d4a50973de26223...         type2   
5   b96eee9a57d57b6fcf3d9326ae8372bd196706da93caa4...         type1   
6   8a2da88b5859a2cf5635a3ca35da1d3456fadd4272f9e3...         type1   
7   73d26fe0f67c6731617902c43fba936e5ff55c36bd12ff...         type1   
8   db17945b2ce9557cb456172f94d43ea170d045d92282b8...         type1   
9   1476f11455ec5e9c2d2d6392545178d8c6184257134b45...         type2   
10  ebf4dee584d8b5fd1958b41791182633f6230fb9592499...         type1   
11  517c7bfb2dca210564375cc9aa13fd659a3da9b2eb7cfb...         type1   
12  ccd5a490a57c4297b170659b411443c97cbffd99be76f6...         type1   
13  beaae9d14c5fc2b3a40345f899683d1f0094c303eb2b01...         type1   
14  8d13238cd8fda0ef080078ee00f3c7d27177db1789e905...         type1   
15  e7b770ba3dc2b600f7bd1c529f88177afec0aca3b07578...         type1   
16  6b403d3818f30f2b4f190e8202b8d2044b10538c6b66ee...         type1   
17  dea5c80db34289d04123421fa63d5f0cade3387d0b54cf...         type2   
18  d7b530ea1617d17033ced725d6128c3273c11c30551286...         type2   
19  1e870dbefc2b6027f627fdc3217d39f891ad6a01e8f8c8...         type2   

    currentAge  diagnosisAge  yearsLivingWithDiabetes  basal.count  \
0           59            20                       38          NaN   
1           65            45                       19          NaN   
2           32            15                       17       2712.0   
3           47            14                       32          NaN   
4           62            53                        8          NaN   
5            4             4                        0          NaN   
6           26             7                       19          NaN   
7           18             8                       10          NaN   
8           45             5                       40          NaN   
9           39            23                       16          NaN   
10           6             2                        3        490.0   
11          24            13                       10        744.0   
12          45            30                       15        792.0   
13          35            25                        9        270.0   
14          51            12                       39        785.0   
15          40            10                       29          NaN   
16          32             4                       28        182.0   
17          60            40                       20          NaN   
18          58            54                        4          NaN   
19          75            37                       37          NaN   

    bolus.count  cbg.count  deviceEvent.count  pumpSettings.count  \
0           NaN        NaN                NaN                 NaN   
1           NaN        NaN                NaN                 NaN   
2        1268.0    29811.0              637.0                24.0   
3           NaN    10279.0                2.0                 NaN   
4           NaN        NaN                NaN                 NaN   
5           NaN        NaN                NaN                 NaN   
6           NaN     4409.0                1.0                 NaN   
7           NaN    14149.0                NaN                 NaN   
8           NaN     5443.0                NaN                 NaN   
9           NaN      344.0                NaN                 NaN   
10        521.0   219077.0             1919.0                 3.0   
11        462.0    23323.0               48.0                 7.0   
12        458.0    11520.0              306.0                 5.0   
13        492.0    53079.0              107.0                 1.0   
14       1643.0   743973.0             7799.0    

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386 entries, 0 to 385
Data columns (total 26 columns):
hashID                        386 non-null object
diagnosisType                 386 non-null object
currentAge                    386 non-null int64
diagnosisAge                  386 non-null int64
yearsLivingWithDiabetes       386 non-null int64
basal.count                   170 non-null float64
bolus.count                   170 non-null float64
cbg.count                     288 non-null float64
deviceEvent.count             229 non-null float64
pumpSettings.count            170 non-null float64
smbg.count                    272 non-null float64
upload.count                  386 non-null int64
wizard.count                  169 non-null float64
cbg.nUniqueDeviceIds          288 non-null float64
bolus.nUniqueDeviceIds        170 non-null float64
smbg.nUniqueDeviceIds         272 non-null float64
basal.nDaysWithData           170 non-null float64
bolus.nDaysWithData           170 non-

In [6]:
data.isnull().sum()

hashID                          0
diagnosisType                   0
currentAge                      0
diagnosisAge                    0
yearsLivingWithDiabetes         0
basal.count                   216
bolus.count                   216
cbg.count                      98
deviceEvent.count             157
pumpSettings.count            216
smbg.count                    114
upload.count                    0
wizard.count                  217
cbg.nUniqueDeviceIds           98
bolus.nUniqueDeviceIds        216
smbg.nUniqueDeviceIds         114
basal.nDaysWithData           216
bolus.nDaysWithData           216
cbg.nDaysWithData              98
deviceEvent.nDaysWithData     157
pumpSettings.nDaysWithData    216
smbg.nDaysWithData            114
upload.nDaysWithData            0
wizard.nDaysWithData          217
cgmSettings.count             321
cgmSettings.nDaysWithData     321
dtype: int64

In [34]:
data.columns

Index(['hashID', 'diagnosisType', 'currentAge', 'diagnosisAge',
       'yearsLivingWithDiabetes', 'basal.count', 'bolus.count', 'cbg.count',
       'deviceEvent.count', 'pumpSettings.count', 'smbg.count', 'upload.count',
       'wizard.count', 'cbg.nUniqueDeviceIds', 'bolus.nUniqueDeviceIds',
       'smbg.nUniqueDeviceIds', 'basal.nDaysWithData', 'bolus.nDaysWithData',
       'cbg.nDaysWithData', 'deviceEvent.nDaysWithData',
       'pumpSettings.nDaysWithData', 'smbg.nDaysWithData',
       'upload.nDaysWithData', 'wizard.nDaysWithData', 'cgmSettings.count',
       'cgmSettings.nDaysWithData'],
      dtype='object')

Intial Obserservations:
- Lot's of missing values
- 'cgmSettings.count' and 'cgmSettings.nDaysWithData missing the most 
- data only has 386 rows. Some features are missing 83% of their data.
    - so just indiscrimantly dropping features is not an 
    - modeling can be based on particular subgroups
    
- Hash ID is an anonymizing feature 
- Basil count: 
    - broken down into more categories  : have records, dont have records. If they have records, in new columns the record will have
- All fields  that are missing over 80% we will break down
    

In [27]:
data.head()

hashID diagnosisType  \
0  92c2847d30b6514ac8a6385f04b3bf3c92ee87a8b89cb8...         type2   
1  6b720ecfc94a90f7c75a39f80a97eac27969ca95a2e6fb...         type2   
2  e0ef840bd07fc79f10b8e21a7f6f3d53ef3b7594d94ec0...         type1   
3  8974377383af45404d560543b58749b240c78885ca0710...         type1   
4  f92417af60cc16f254600f310df4d15d4a50973de26223...         type2   

   currentAge  diagnosisAge  yearsLivingWithDiabetes  basal.count  \
0          59            20                       38          NaN   
1          65            45                       19          NaN   
2          32            15                       17       2712.0   
3          47            14                       32          NaN   
4          62            53                        8          NaN   

   bolus.count  cbg.count  deviceEvent.count  pumpSettings.count  \
0          NaN        NaN                NaN                 NaN   
1          NaN        NaN                NaN                 NaN   
2       1268.0    29811.0              637.0                24.0   
3          NaN    10279.0                2.0                 NaN   
4          NaN        NaN                NaN                 NaN   

             ...              basal.nDaysWithData  bolus.nDaysWithData  \
0            ...                              NaN                  NaN   
1            ...                              NaN                  NaN   
2            ...                            213.0                212.0   
3            ...                              NaN                  NaN   
4            ...                              NaN                  NaN   

   cbg.nDaysWithData  deviceEvent.nDaysWithData  pumpSettings.nDaysWithData  \
0                NaN                        NaN                         NaN   
1                NaN                        NaN                         NaN   
2              110.0                      150.0                        23.0   
3              112.0                        2.0                         NaN   
4                NaN                        NaN                         NaN   

   smbg.nDaysWithData  upload.nDaysWithData  wizard.nDaysWithData  \
0                48.0                     1                   NaN   
1                41.0                    14                   NaN   
2               213.0                     5                 212.0   
3                16.0                     1                   NaN   
4               400.0                     1                   NaN   

   cgmSettings.count  cgmSettings.nDaysWithData  
0                NaN                        NaN  
1                NaN                        NaN  
2                NaN                        NaN  
3                NaN                        NaN  
4                NaN                        NaN  

[5 rows x 26 columns]

In [47]:
data['hashID'].unique().shape

(386,)

In [39]:
type_1 = data[data['diagnosisType'] == 'type1']

In [40]:
type_2 = data[data['diagnosisType'] == 'type2']

In [41]:
plt.scatter(type_1['currentAge'])

TypeError: scatter() missing 1 required positional argument: 'y'

In [30]:
data[data['diagnosisType'] == 'type1'].count()

hashID                        328
diagnosisType                 328
currentAge                    328
diagnosisAge                  328
yearsLivingWithDiabetes       328
basal.count                   163
bolus.count                   163
cbg.count                     254
deviceEvent.count             206
pumpSettings.count            163
smbg.count                    236
upload.count                  328
wizard.count                  162
cbg.nUniqueDeviceIds          254
bolus.nUniqueDeviceIds        163
smbg.nUniqueDeviceIds         236
basal.nDaysWithData           163
bolus.nDaysWithData           163
cbg.nDaysWithData             254
deviceEvent.nDaysWithData     206
pumpSettings.nDaysWithData    163
smbg.nDaysWithData            236
upload.nDaysWithData          328
wizard.nDaysWithData          162
cgmSettings.count              64
cgmSettings.nDaysWithData      64
dtype: int64

In [31]:
data[data['diagnosisType'] == 'type2'].count() # most of basil count is missing for them 

hashID                        58
diagnosisType                 58
currentAge                    58
diagnosisAge                  58
yearsLivingWithDiabetes       58
basal.count                    7
bolus.count                    7
cbg.count                     34
deviceEvent.count             23
pumpSettings.count             7
smbg.count                    36
upload.count                  58
wizard.count                   7
cbg.nUniqueDeviceIds          34
bolus.nUniqueDeviceIds         7
smbg.nUniqueDeviceIds         36
basal.nDaysWithData            7
bolus.nDaysWithData            7
cbg.nDaysWithData             34
deviceEvent.nDaysWithData     23
pumpSettings.nDaysWithData     7
smbg.nDaysWithData            36
upload.nDaysWithData          58
wizard.nDaysWithData           7
cgmSettings.count              1
cgmSettings.nDaysWithData      1
dtype: int64

- We have very unbalanced classes: ~ 6:1 ratio for Type1. Type1 is dominate class. 
